File: NTTU_SVPCC/2023NTTU_SVPCC_2/PE.py
# while True:
#     try:
#         n=int(input())
#         words=[]  
#         no_vowel=[]
#         for i in range(n):
#             words.append(input())
#             t=""
#             for j in range(len(words[i])):
#                 if words[i][j]!="A" and words[i][j]!="E" and words[i][j]!="I" and words[i][j]!="O" and words[i][j]!="U":
#                     t+=words[i][j]
#             no_vowel.append(t)
#         message=str(input())
#         temp=""
#         for i in range(len(message)):
#             temp+=message[i]
#             c=no_vowel.count(temp)
#             if c!=0:
#                 ind=no_vowel.index(temp)
#                 if i!=len(message)-1:
#                     print(words[ind],end=" ")
#                 else:
#                     print(words[ind])
#                 temp=""
#     except EOFError:
#         break
def get_no_vowel_words(words):
    no_vowel_words = set()
    for word in words:
        no_vowel = ''.join(c for c in word if c not in 'AEIOU')
        no_vowel_words.add(no_vowel)
    return no_vowel_words

def main():
    n = int(input())
    words = []
    for _ in range(n):
        words.append(input())

    no_vowel_words = get_no_vowel_words(words)

    message = input()
    temp = ""
    for char in message:
        temp += char
        if temp in no_vowel_words:
            ind = [i for i, word in enumerate(words) if ''.join(c for c in word if c not in 'AEIOU') == temp]
            print(words[ind[0]], end=" ")
            temp = ""

if __name__ == "__main__":
    main()

File: NTTU_SVPCC/2023NTTU_SVPCC_2/PE.cpp
#include <iostream>
#include <vector>
#include <unordered_map>
using namespace std;

int main() {
    int n;
    while (cin >> n) {
        vector<string> words(n);
        vector<string> no_vowel;
        unordered_map<string, int> no_vowel_words;

        for (int i = 0; i < n; i++) {
            cin >> words[i];
            string t = "";
            for (int j = 0; j < words[i].length(); j++) {
                if (words[i][j] != 'A' && words[i][j] != 'E' && words[i][j] != 'I' && words[i][j] != 'O' && words[i][j] != 'U') {
                    t += words[i][j];
                }
            }
            no_vowel.push_back(t);
            no_vowel_words[t] = i;
        }

        string message;
        cin.ignore(); // Ignore the newline character after n input
        getline(cin, message);

        string temp = "";
        for (int i = 0; i < message.length(); i++) {
            temp += message[i];
            if (no_vowel_words.count(temp)) {
                int ind = no_vowel_words[temp];
                if (i != message.length() - 1) {
                    cout << words[ind] << " ";
                } else {
                    cout << words[ind] << endl;
                }
                temp = "";
            }
        }
    }

    return 0;
}
